modules/base/src/algorithm/mesh/meshclipping.cpp

```diff
@@ -0,0 +1,572 @@
+        loop.push_back(edges.back()[0]);
+        loop.push_back(edges.back()[1]);
+        edges.pop_back();
+
+        while (!edges.empty()) {
+            const auto [it, index] = findMatch(edges, loop.back());
+            if (it != edges.end()) {
+                if (glm::all(glm::equal(positions[index], positions[loop.front()], eps))) {
+                    edges.erase(it);
+                    break;
+                } else {
+                    loop.push_back(index);
+                    edges.erase(it);
+                }
+            } else {
+                LogWarnCustom(
+                    "MeshClipping",
+                    "Found edge, that is not connected to any other edge. This could mean, the "
+                    "clipped mesh was not manifold.");
+                break;
+            }
+        }
+    }
+    return loops;
+}
+
+std::vector<float> barycentricInsidePolygon(vec2 v, const std::vector<vec2>& vis) {
+
+    std::vector<vec2> sis;
+    std::transform(vis.begin(), vis.end(), std::back_inserter(sis),
+                   [v](const vec2& vi) { return vi - v; });
+
+    std::vector<float> weights(vis.size(), 0.0f);
+    std::vector<float> tana(vis.size(), 0.0f);
+
+    for (size_t i = 0; i < vis.size(); ++i) {
+        const size_t j = (i + 1) % vis.size();
+        const auto ri = glm::length(sis[i]);
+        const auto Aij = glm::determinant(mat2{sis[i], sis[j]});
+        const auto Dij = glm::dot(sis[i], sis[j]);
+        if (util::almostEqual(ri, 0.0f)) {
+            weights[i] = 1.0f;
+            return weights;
+        }
+        const auto rj = glm::length(sis[j]);
+
+        if (util::almostEqual(Aij, 0.0f)) {
+            if (Dij < 0.0f) {
+                weights[i] = ri / (ri + rj);
+                weights[j] = rj / (ri + rj);
+                return weights;
+            }
+        } else {
+            tana[i] = (rj - Dij / ri) / Aij;
+        }
+    }
+
+    float total = 0.0f;
+    for (size_t i = 0; i < vis.size(); ++i) {
+        const auto prev = (i - 1 + vis.size()) % vis.size();
+        weights[i] = tana[prev] + tana[i];
+        total += weights[i];
+    }
+    std::transform(weights.begin(), weights.end(), weights.begin(),
+                   [total](float w) { return w / total; });
+
+    return weights;
+}
+
+void capHoles(std::vector<glm::u32vec2>& edges, vec3 normal, const std::vector<vec3>& positions,
+              std::vector<std::uint32_t>& indices,
+              const InterpolateFunctor& addInterpolatedVertex) {
+
+    constexpr float relError = 0.000001f;
+    const auto eps =
+        relError * std::accumulate(edges.begin(), edges.end(), 0.0f, [&](float m, glm::u32vec2 e) {
+            return glm::max(m, glm::max(glm::compMax(glm::abs(positions[e[0]])),
+                                        glm::compMax(glm::abs(positions[e[1]]))));
+        });
+
+    removeDuplicateEdges(edges, positions, eps);
+    const auto loops = gatherLoops(edges, positions, eps);
```

Comment:
Did you notice any further problems because of the epsilon?

Reply:
Not in the meshes I've tried with, but I have not tested with very complex ones...